forem/forem
# Allow users to pause gifs in a post

## What type of PR is this? (check all applicable)


## Description

This PR adds the ability for users to pause playback of any images with attribute `data-animated="true"`. 

At the moment, detection of animated images is behind the feature flag `:detect_animated_images`, which I believe is only enabled on DEV. Therefore this change will only affect DEV at the moment.

We recently discovered that our code to detect animated images will only pick up those with `.gif` file extensions, and because of the way we process gifs in posts in cloud-hosted forems (converting `.gif` to  `.webp`), further backend work would be needed to roll this feature out further (beyond just enabling the feature flag elsewhere).

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/188

## QA Instructions, Screenshots, Recordings

To test this fully, you will need to toggle your system "reduced motion" preferences, or emulate the different settings in dev tools (probably the simplest approach). In chrome dev tools you can emulate 'reduced motion' by bringing up the command palette with `cmd + shift + p` and searching for 'prefers-reduced-motion'. You'll need to refresh the page after toggling the emulation.

- Before testing, enable the feature flag by running `FeatureFlag.enable(:detect_animated_images)` in the console
- Publish a post that links out to a gif (e.g. from giphy)
- When you first publish the post, the gifs will not be pausable, this is because the animated image detection happens async behind the scenes
- Refresh the article page
- Gifs should load in, and a toggle playback button should appear in the bottom-right corner
- If you do not have "reduced motion" set: Playback of the gif should resume once the 'pausable' version has loaded in
- If you do have "reduced motion" set: Gifs should be paused as they load in (animation may be briefly visible)
- Clicking the button should toggle the playback 
- The button should be clickable by either mouse or keyboard

NB: when a gif is paused, the first frame of the gif shows (not whatever frame you clicked pause on)



https://user-images.githubusercontent.com/20773163/151334286-ece75a38-2a53-499b-a6e5-7ac58f66a8ac.mp4


### UI accessibility concerns?

This helps address an accessibility concern where animated images can trigger adverse reactions in users.

## Added/updated tests?


- [X] No, and this is why: I'm not totally sure how best to test this. The underlying play/pause functionality comes from a library which has its own testing.


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
